Integrations/integration-MaxMind_GeoIP2.yml

```diff
@@ -0,0 +1,293 @@
+description: Enriches IP addresses
+detaileddescription: 'The MaxMind GeoIP2 integration allows you to query the MaxMind
+  API service and retrieve a JSON of all details.'
+configuration:
+- display: API Key
+  name: apikey
+  defaultvalue: ""
+  type: 0
+  required: true
+- display: Account ID
+  name: account
+  defaultvalue: ""
+  type: 0
+  required: true
+- display: Use system proxy
+  name: proxy
+  defaultvalue: ""
+  type: 8
+  required: false
+- display: Trust any certificate (unsecure)
+  name: insecure
+  defaultvalue: ""
+  type: 8
+  required: false
+- display: Service Version
+  name: mode
+  defaultvalue: city
+  type: 15
+  required: true
+  options:
+  - country
+  - city
+  - insights
+- display: Base URL
+  name: url
+  defaultvalue: https://geoip.maxmind.com/geoip/
+  type: 0
+  required: true
+script:
+  script: |
+    import requests
+    from collections import defaultdict
+    from requests.auth import HTTPBasicAuth
+    # disable insecure warnings
+    requests.packages.urllib3.disable_warnings()
+
+
+    '''GLOBAL VARS'''
+    BASE_URL = demisto.params().get('url')
+    APIKEY = demisto.params().get('apikey')
+    ACCOUNT_ID = demisto.params().get('account')
+    MODE = demisto.params().get('mode')
+    INSECURE = demisto.params().get('insecure')
+    PROXY = demisto.params().get('proxy')
+    API_VERSION = 'v2.1'
+
+    if not demisto.params().get('proxy', False):
+        del os.environ['HTTP_PROXY']
+        del os.environ['HTTPS_PROXY']
+        del os.environ['http_proxy']
+        del os.environ['https_proxy']
+
+    HR_HEADERS = [
+        'IP',
+        'Domain',
+        'ASN',
+        'Organization',
+        'ISP',
+        'Location',
+        'Accuracy Radius',
+        'User Type',
+        'Continent',
+        'ISO Code',
+        'Country',
+        'Registered Country',
+        'TimeZone',
+        'City',
+        'Subdivision',
+        'Is TOR Exit Node',
+        'Is Hosting Provider',
+        'Is Anonymous']
+
+    HEADERS = {
+        'Content-Type': 'application/json',
+        'Accept': 'application/json'
+    }
+
+
+    '''HELPER FUNCTIONS'''
+    def http_request(query):
+        r = requests.request(
+            'GET',
+            BASE_URL + API_VERSION + '/' + MODE +'/'+ query,
+            headers=HEADERS,
+            verify=INSECURE,
+            auth=HTTPBasicAuth(ACCOUNT_ID, APIKEY)
+        )
+        if r.status_code is not 200:
+            return_error('Error in API call to MaxMind, got status code - {} and a reason: {}'.format(r.status_code, r.reason))
+        return r
+
+
+    def format_results(res_json):
+        hr = defaultdict()
+        ec = defaultdict()
+        maxmind_ec = defaultdict(lambda: defaultdict(int))
+        ip_ec = defaultdict(lambda: defaultdict(int))
+        if 'continent' in res_json:
```

Comment:
this is not about this lin specifically, but about the following context creation.
Is it really the case that you are not sure if you get continent/city/location/registered_country[..] and so on?
are those nested ifs necessary?
because if they will always exist as a group or won't exist at all, we can lose most of the ifs and make it much more readable context creation.
and create the sub dictionaries statically

Reply:
Unfortunately yes. In many cases, they will report only partial information and when that happens, it throws a key error.

The response that they return, it is not standardised in the sense that a call for a specific mode will always return the country, city, or location results. In my testing, the API returns what ever information they have and exclude certain values depending on the mode.

For example, if you submit an IP using the country mode, you'll receive the country (if it exists) and not city or registered country since they are only delivered for different modes.